src/monitor.cpp

```diff
@@ -160,108 +160,139 @@ complex<double> fields_chunk::get_field(component c, const ivec &iloc) const {
-  if (is_mine())
-    return s->chi1inv[c][d] ? s->chi1inv[c][d][gv.index(c, iloc)]
-                            : (d == component_direction(c) ? 1.0 : 0);
-  return 0.0;
+double fields_chunk::get_chi1inv(component c, direction d, const ivec &iloc, double omega) const {
+  return s->get_chi1inv(c, d, iloc, omega);
 }
 
-double fields::get_chi1inv(component c, direction d, const vec &loc, bool parallel) const {
+double fields::get_chi1inv(component c, direction d, const vec &loc, double omega, bool parallel) const {
   ivec ilocs[8];
   double w[8], res = 0.0;
   gv.interpolate(c, loc, ilocs, w);
   for (int argh = 0; argh < 8 && w[argh] != 0; argh++)
-    res += w[argh] * get_chi1inv(c, d, ilocs[argh], false);
+    res += w[argh] * get_chi1inv(c, d, ilocs[argh], omega, false);
   return parallel ? sum_to_all(res) : res;
 }
 
-double fields::get_eps(const vec &loc) const {
+double fields::get_eps(const vec &loc, double omega) const {
   double tr = 0;
   int nc = 0;
   FOR_ELECTRIC_COMPONENTS(c) {
     if (gv.has_field(c)) {
-      tr += get_chi1inv(c, component_direction(c), loc, false);
+      tr += get_chi1inv(c, component_direction(c), loc, omega, false);
       ++nc;
     }
   }
   return nc / sum_to_all(tr);
 }
 
-double fields::get_mu(const vec &loc) const {
+double fields::get_mu(const vec &loc, double omega) const {
   double tr = 0;
   int nc = 0;
   FOR_MAGNETIC_COMPONENTS(c) {
     if (gv.has_field(c)) {
-      tr += get_chi1inv(c, component_direction(c), loc, false);
+      tr += get_chi1inv(c, component_direction(c), loc, omega, false);
       ++nc;
     }
   }
   return nc / sum_to_all(tr);
 }
 
-double structure::get_chi1inv(component c, direction d, const ivec &origloc, bool parallel) const {
+double structure::get_chi1inv(component c, direction d, const ivec &origloc, double omega, bool parallel) const {
   ivec iloc = origloc;
   for (int sn = 0; sn < S.multiplicity(); sn++)
     for (int i = 0; i < num_chunks; i++)
       if (chunks[i]->gv.owns(S.transform(iloc, sn))) {
         signed_direction ds = S.transform(d, sn);
-        double val = chunks[i]->get_chi1inv(S.transform(c, sn), ds.d, S.transform(iloc, sn)) *
+        double val = chunks[i]->get_chi1inv(S.transform(c, sn), ds.d, S.transform(iloc, sn), omega) *
                      (ds.flipped ^ S.transform(component_direction(c), sn).flipped ? -1 : 1);
         return parallel ? sum_to_all(val) : val;
       }
   return 0.0;
 }
+// Useful if you already know the exact location of the point you are interested in
+// (i.e. you know what idx should be). This is used with the get_array() routines.
+double structure_chunk::get_chi1inv_at_pt(component c, direction d, int idx, double omega) const {
+  double res = 0.0;
+  if (is_mine()){
+    res =
+        chi1inv[c][d] ? chi1inv[c][d][idx] : (d == component_direction(c) ? 1.0 : 0);
+    if (res != 0){
+      // Get instaneous dielectric (epsilon)
+      std::complex<double> eps(1.0 / res,0.0);
```

Comment:
This seems problematic for off-diagonal components, or in general for anisotropic materials, since you need a matrix inverse in that case.

Reply:
Addressed in 6eac066